### ISSUE TITLE
False positive on collapsible_if with mutating function

### ISSUE THREAD

**Issue by jean-airoldie** [rust-lang/rust-clippy#4261]
With `clippy 0.0.212 (265318d 2019-05-17)`, I get a false positive for the collapsible_if lint.

```
warning: this if statement can be collapsed
   --> src/lib.rs:247:9
    |
247 | /         if target.contains("msvc") {
248 | |             if rename_libzmq_in_dir(&lib_dir, "zmq.lib").is_err() {
249 | |                 panic!("unable to find compiled `libzmq` lib");
250 | |             }
251 | |         }
    | |_________^
    |
    = note: #[warn(clippy::collapsible_if)] on by default
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#collapsible_if
help: try
    |
247 |         if target.contains("msvc") && rename_libzmq_in_dir(&lib_dir, "zmq.lib").is_err() {
248 |     panic!("unable to find compiled `libzmq` lib");
249 | }
    |
```

In this case the function `rename_libzmq_in_dir` mutates state which means that the suggestion is not equivalent. Indeed, using the suggestion would run the function every time, as opposed to only when the first condition is met.

**Comment by llogiq** [comment]
No. `&&` is *short-circuiting*, which means that the right-hand side expression is only evaluated if the left-hand side evaluates to `true`.
